app_dev.yaml

```diff
@@ -17,7 +17,11 @@ inbound_services:
 
 default_expiration: "10m"
 handlers:
-# PRODUCTION STATIC
+# PRODUCTION STATIC START
```

Comment:
Wouldn't these rules also be captured in the dev build? I don't see any code that excludes them from it.

Reply:
There is no app.yaml generation for dev mode. This file is directly used.

Reply:
Done.

Reply:
I don't see the explanation (comments/documentation) I asked for in my previous comment (I see the comments above but they aren't related to my question)...

Reply:
Oh, I added it [here](https://github.com/oppia/oppia/pull/8488/files#diff-191be61dc824545e1299bcaae58a8b58R1273) when `app.yaml` is being generated. Should I add it in `app_dev.yaml` too?

Reply:
Ah -- yes, definitely, please do so. 

But I'm also not sure it addresses the question. My specific question is: for handlers like third_party/static on line 27, this file suggests that we should be excluding them from local dev mode (because they are in the prod section). But we don't seem to be actually excluding them, because, for local dev mode, we just use app_dev.yaml directly, right? (Please let me know if I'm misunderstanding something.)

Reply:
Yes, I think your concern is right here. We are not excluding anything for a dev build. I am not sure about why this pattern exists initially. I think for local builds, we do not have a major concern since it is only used for testing. So, maybe the initial design was made with the idea of optimising build for deployment.

@vojtechjelinek @kevinlee12 can you please confirm the same?

Also, on further thoughts,  I think this shouldn't be dev & prod. It should be deploy & non-deploy build because even if we do a prod build locally, we will need folders like `assets/scripts` for e2e tests which are not required while deploying.

Should I refactor my PR for this change as well as add a new method for generating `app.yaml` for non-deploy mode?

Reply:
Would having to generate app.yaml for non-deploy mode cause any problems when contributors are running a local dev server?

Reply:
@ankita240796 I'm unfamiliar with the history of the mentioned files, so I'd have to defer.

Reply:
The app_dev.yaml was initially introduced to rewrite the static paths and to serve the HTML files for /build folder, it is both for deployment and also the local prod because we want to have as little differences between local prod and deploy as possible.

Reply:
> Would having to generate app.yaml for non-deploy mode cause any problems when developers are running a local dev server? What if the developer updates the app.yaml file (or the original app_dev.yaml file) during development?

I am not sure about this but I think yes this can cause issues if the developer tries to modify app_dev.yaml during deployment expecting a new directory to be uploaded. I think we should let the pattern for local builds be the same but modify the pattern for deploy build (i.e exclude non-deploy handlers from being uploaded.)

> The app_dev.yaml was initially introduced to rewrite the static paths and to serve the HTML files for /build folder, it is both for deployment and also the local prod because we want to have as little differences between local prod and deploy as possible.

But shouldn't we have some difference between local prod & deploy, because in local prod we upload folders like assets/scripts for testing but they aren't needed in deploy mode.

Reply:
> But shouldn't we have some difference between local prod & deploy, because in local prod we upload folders like assets/scripts for testing but they aren't needed in deploy mode.

Sure, I'm just stating that the files are now used to differ between dev and prod, the difference between local and deploy might be needed to reduce the number of files but I don't think that this is really a thing that we want but moreover a thing that we need in order to have less than 10000 files on the deploy server.

Reply:
Understood, Thanks for the clarification!

Reply:
@seanlip, Here is what I think we should do:

- Do not make any changes for app.yaml for generation for local builds i.e. it should be same as `app_dev.yaml` for local dev and for local prod, it should be generated using `generate_app_yaml` function in build script without removal of any section.
- Since our main concern is to avoid uploading more than 10000 files on the deploy server, we should only remove stuff from `app_dev.yaml` in deploy mode.

What do you think?

Reply:
I think the only thing that reduces the number of files (below the 10k limit) is skip_files. That doesn't seem to be what we are referring to here, right?

If we're talking about the handlers section, it's more a question about what endpoints we want to expose in the local dev, local prod and deployed contexts, and what handlers we want these endpoints to map to in these cases. I'm not sure the action plan you proposed addresses that?

Reply:
Ok, I agree with you that changing the handlers won't affect the file limit a lot.

So, if we want to decide the endpoints we want to expose, isn't it still ok to expose all endpoints for a local build and only the required ones for deployment?

Reply:
To be more specific: A local build is just done for testing features/any tests locally, so we should not worry about which handlers we expose. On the other hand, when we deploy, we serve content on our prod website and we should avoid exposing any handlers which are not used.

Reply:
OK, I'm fine with that. Maybe add the rationale as a comment to app_dev.yaml.

Please refactor accordingly; in particular, make sure that the labels are not misleading (i.e. don't say something is "prod only" or "dev only" when it isn't).

Reply:
Done.